instrumentation/apache-httpasyncclient-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpasyncclient/ApacheHttpClientRequest.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.apachehttpasyncclient;
+
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.hc.core5.http.EntityDetails;
+import org.apache.hc.core5.http.Header;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpResponse;
+import org.apache.hc.core5.http.ProtocolVersion;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class ApacheHttpClientRequest {
+
+  private static final Logger logger = LoggerFactory.getLogger(ApacheHttpClientRequest.class);
+
+  @Nullable private final URI uri;
+
+  private final HttpRequest delegate;
+  private final EntityDetails entityDetails;
+
+  public ApacheHttpClientRequest(HttpRequest httpRequest, EntityDetails entityDetails) {
+    uri = getUri(httpRequest);
+    this.entityDetails = entityDetails;
+    this.delegate = httpRequest;
+  }
+
+  public List<String> getHeader(String name) {
+    return headersToList(delegate.getHeaders(name));
+  }
+
+  // minimize memory overhead by not using streams
+  static List<String> headersToList(Header[] headers) {
+    if (headers.length == 0) {
+      return Collections.emptyList();
+    }
+    List<String> headersList = new ArrayList<>(headers.length);
+    for (int i = 0; i < headers.length; ++i) {
+      headersList.add(headers[i].getValue());
+    }
+    return headersList;
+  }
+
+  public Long requestContentLength() {
+    return entityDetails != null ? entityDetails.getContentLength() : null;
+  }
+
+  public void setHeader(String name, String value) {
+    delegate.setHeader(name, value);
+  }
+
+  public String getMethod() {
+    return delegate.getMethod();
+  }
+
+  public String getUrl() {
+    return uri != null ? uri.toString() : null;
+  }
+
+  public String getFlavor(HttpResponse response) {
```

Comment:
Nit
```suggestion
  public String getFlavor(@Nullable HttpResponse response) {
```

Reply:
This was actually done for interim purpose, actually if we see this is kind of a static method or more related to response and not the request, what do you suggest I should do here. This should ideally not be part of the AsyncHttpClientRequest now as it is now not a request property.

Reply:
I will need to add a logger separately in the AttributesGetter as well that should be fine right?